---
title: "GO enrichment"
output: html_notebook
---
## Gene expression analysis

- detection of down- and upregulated genes
- using code from Inga for the analysis

```{r,  echo=FALSE}
library(stringr)
library(DESeq2)
library(dplyr)

source('~/DataScienceProject/RNA-seq/differentialexpression_function.R')
```

```{r}
# import data
rnaseqdata <- read.table("~/Dokumente/GSE153873_summary_count.star.txt", sep = "\t",header = TRUE,row.names = 1)
colnames(rnaseqdata) <- paste0(data.frame(str_split(colnames(rnaseqdata),'[.]'))[3,],data.frame(str_split(colnames(rnaseqdata),'[.]'))[1,])
#rnaseq_ercc <- read.table("RNAseq/GSE153873_summary_count.ercc.txt", sep = "\t")
dim(rnaseqdata)
# 27135 different genes and 30 patients

# Groups
AD <- dplyr::select(rnaseqdata, starts_with("AD"))
young <- dplyr::select(rnaseqdata, starts_with("Young"))
old <- dplyr::select(rnaseqdata, starts_with("Old"))

# new order of patient
rnaseqdata <- data.frame(AD,old,young)

# DIFFERENTIAL EXPRESSION
res <- diffexpression(old,AD,alpha=0.05,tidy=FALSE)
summary(res)

#421 genes with significant upregulation in AD, while 434 had significant downregulation -> 855 significant up/downregulated


pvalues <- p.adjust(res$pvalue, method = "BY")
sum(pvalues<0.05,na.rm = T)

#434 downregulated, 421 upregulated

data<- data_frame(log2FoldChange=res$log2FoldChange, padj= res$padj, Yekutielipval=pvalues,pval=res$pvalue, genes=rownames(res))
data<- data %>%
  mutate(gene_type = case_when(log2FoldChange > 0 & padj < 0.05 ~ "upregulated",
                               log2FoldChange < 0 & padj < 0.05 ~ "downregulated",
                               TRUE ~ "not significant"))   
print(table(data$gene_type))

```

## GO enrichment

- using clusterprofiler and enrichplot 
- first step is to create two lists of the gene subsets (down/up-regulated) containing gene names and their correpsonding log2 foldchanges 


```{r}
#up and downregulated genes, ordered by p_adjust
upregulated <- data[data$gene_type=="upregulated",]
upregulated <- upregulated[order(upregulated$padj),]
downregulated <- data[data$gene_type=="downregulated",]
downregulated <- downregulated[order(downregulated$padj),]

# creating list for enrichment analysis
# containing gene name and log2 fold change 
upregulated_list <- 2^(upregulated$log2FoldChange) 
names(upregulated_list)<- as.character(upregulated$genes)
upregulated_list <- sort(upregulated_list, decreasing=TRUE)

downregulated_list <- 2^(downregulated$log2FoldChange) 
names(downregulated_list)<- as.character(downregulated$genes)
downregulated_list <- sort(downregulated_list, decreasing=TRUE)

saveRDS(downregulated_list, file = "~/R_downregulated")
saveRDS(upregulated_list, file = "~/R_upregulated")

```

- using function form clusterprofler to enrich GO terms : biological process, cellular component and molecular function 

```{r}
library(enrichplot)
library(DOSE)
library(org.Hs.eg.db)
library(clusterProfiler)

downregulated_list<- readRDS("~/R_downregulated")
upregulated_list <- readRDS("~R_upregulated")
### upregualted GO terms
up_BPs <- gseGO( upregulated_list, 
                 keyType = "SYMBOL",
                 OrgDb = org.Hs.eg.db, 
                 ont = "BP",
                 pvalueCutoff = 0.05,
                 nPermSimple = 10000 )

up_CCs <- gseGO( upregulated_list,
                 keyType = "SYMBOL",
                 OrgDb = org.Hs.eg.db, 
                 ont = "CC",
                 pvalueCutoff = 0.05,
                 nPermSimple = 10000 )

up_MFs <- gseGO( upregulated_list, 
                 keyType = "SYMBOL",
                 OrgDb = org.Hs.eg.db, 
                 ont = "MF",
                 pvalueCutoff = 0.05,
                 nPermSimple = 10000 )

## downregulated GO terms
down_BPs <- gseGO( downregulated_list, 
                   keyType = "SYMBOL",
                 OrgDb = org.Hs.eg.db, 
                 ont = "BP",
                 pvalueCutoff = 0.05,
                 nPermSimple = 10000 )

down_CCs <- gseGO( downregulated_list, 
                   keyType = "SYMBOL",
                 OrgDb = org.Hs.eg.db, 
                 ont = "CC",
                 pvalueCutoff = 0.05,
                 nPermSimple = 10000 )

down_MFs <- gseGO( downregulated_list, 
                   keyType = "SYMBOL",
                 OrgDb = org.Hs.eg.db, 
                 ont = "MF",
                 pvalueCutoff = 0.05,
                 nPermSimple = 10000 )

```

- lastly, some visualizations : 
  -  dot plot gives a nice overview about the GOs
  - treeplot is similiaar to the dotplot but additionally it clusters the hierach. structured GOs 
  - ridgeplot to visualize the distribution of the log2 fold changes for the signigicantly enriched GOs
```{r}

dotplot(down_BPs, showCategory=10) + ggtitle("Biological Process downregulated GOs")

treeplot(pairwise_termsim(down_BPs), showCategory = 10, hclust_method = "average") + ggtitle( "downregulated Biological Process GO Terms")

ridgeplot(down_BPs, showCategory=10) 
```

## Pathway enrichments 

- enrichment analysis using protein pathways of database KEGG and reactome

- for KEGG, an ID conversion from hgnc symbols to entrez ids is necessary 
- unfortunately, the KEGG enrichment didnt work 

```{r}

convert_ID <- function(genelist){
  ids<-bitr(genelist$genes, fromType = "ALIAS", toType = "ENTREZID", OrgDb= org.Hs.eg.db)
  dedup_ids = ids[!duplicated(ids[c("ALIAS")]),]
  df = right_join(genelist, dedup_ids, by= c("genes" = "ALIAS"))
  gene.list.ID <- (df$log2FoldChange)
  # Name vector with ENTREZ ids
  names(gene.list.ID) <- as.character(df$ENTREZID)
  # sort the list in decreasing order (required for clusterProfiler)
  gene.list.ID = sort(gene.list.ID, decreasing = TRUE)
  return(gene.list.ID)
}

upregulated_list <- convert_ID(upregulated)
downregulated_list <- convert_ID(downregulated)

## not working 
up_kk <- gseKEGG(geneList = upregulated_list,
               organism     = "hsa",
               nPermSimple   = 10000,
               minGSSize    = 1,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               #pAdjustMethod = "BH",
               keyType       = "ncbi-geneid")

```

- instead of KEGG, using reactome pathways 
- visualizing pathway of Citric Acid Cycle, since its related to AD pathologies
```{r}
library(ReactomePA)
down.pathways <- enrichPathway(names(downregulated), organism = "human", pvalueCutoff=0.05, pAdjustMethod = "BH")
up.pathways <-  enrichPathway(names(upregulated), organism = "human", pvalueCutoff=0.05, pAdjustMethod = "BH")

viewPathway("The citric acid (TCA) cycle and respiratory electron transport", foldChange = down.gene.list.ID, readable = TRUE)
               
dotplot(up.pathways, showCategory = 20, title = "Enriched Pathways")

```




